src/os/bluestore/BlueFS.cc

```diff
@@ -827,21 +809,47 @@ void BlueFS::_init_alloc()
         name += devnames[id];
       else
         name += to_string(uintptr_t(this));
+
+      auto reserved = _get_minimal_reserved(id);
+      uint64_t full_size = _get_block_device_size(id);
+      uint64_t free_end = p2align(full_size, alloc_size[id]);
+
+      // Trying to lock the following extents:
+      // [reserved, alloc_size] and [p2align(dev_size, alloc_size), dev_size]
+      // to make all the allocations alligned to alloc_size if possible.
+      // Final decision whether locked head/tail to be maintained is made after
+      // BlueFS replay depending on existing allocations.
+      auto &locked = locked_alloc[id];
+      locked.head_offset = reserved;
+      locked.head_length = p2nphase(reserved, alloc_size[id]);
+      if (free_end < full_size) {
+        locked.tail_offset = free_end;
+        locked.tail_length = full_size - free_end;
+      }
+      string alloc_type = cct->_conf->bluefs_allocator;
       dout(1) << __func__ << " new, id " << id << std::hex
               << ", allocator name " << name
-              << ", allocator type " << cct->_conf->bluefs_allocator
-              << ", capacity 0x" << bdev[id]->get_size()
-              << ", reserved 0x" << block_reserved[id]
-              << ", block size 0x" << alloc_size[id]
-              << ", max alloc size 0x" << super.bluefs_max_alloc_size[id]
+              << ", allocator type " << alloc_type
+              << ", capacity 0x" << full_size
+              << ", reserved 0x" << reserved
+              << ", maybe locked " << locked
+              << ", block size 0x" << bdev[id]->get_block_size()
+              << ", alloc unit 0x" << alloc_size[id]
               << std::dec << dendl;
-      alloc[id] = Allocator::create(cct, cct->_conf->bluefs_allocator,
-				    bdev[id]->get_size(),
-				    super.bluefs_max_alloc_size[id],
+      alloc[id] = Allocator::create(cct,
+                                    alloc_type,
+				    full_size,
+				    bdev[id]->get_block_size(),
 				    name);
-      alloc[id]->init_add_free(
-        block_reserved[id],
-        _get_total(id));
+      alloc[id]->init_add_free(reserved, full_size - reserved);
+      // temporary lock candidate regions to forbid their use
+      bluefs_locked_extents_t res_la = locked.get_merged();
+      if (res_la.head_length) {
+        alloc[id]->init_rm_free(res_la.head_offset, res_la.head_length);
+      }
+      if (res_la.tail_length) {
+        alloc[id]->init_rm_free(res_la.tail_offset, res_la.tail_length);
+      }
```

Comment:
Wouldn't it be better to skip doing this in `_init_alloc()`, and do it in `mkfs()` ?
The only other place we invoke `_init_alloc()` is `mount()`, and we remove these restrictions right before marking allocations.